### PR TITLE
Added ical import

### LINKED ISSUE #46: add external calendar support
It would be great for admins to be able to add external calendars to the main yoursite/calendar view. I suspect just iCal support would be enough for most cases. When clicking into an imported event, it could either reference a webpage or display HTML content. You may need to an option to strip HTML out for security reasons though?

If you wanted to go crazy, you could build something that allows mapping incoming data from a file or other source. Something like reading from a json file periodically and mapping that information to the iCal standard. That might be way beyond a normal persons use case though.

My personal use case is to get [this calendar](http://www.opencritic.com/calendar) displayed in my forums calendar.

### PR DESCRIPTION
**Pull request by fank**
What if done:
- Redesigned admin view to match more material style
- Added option to add icals

... and maybe forgot something

TODO:
- [x] Add cronjob for pulling the icals, to local storage. So calander view can be open faster.
- [ ] Add option to see the added icals @pitaj 
- [ ] Add option to deleted the added icals @pitaj 
- [x] Test everything to verify nothing is messed up

Fixes #46 

Tested icals:
- http://c.opencritic.com/calendar/OpenCritic.ics
- http://www.chrisgregori.co.uk/playdate/export.php

image   
image   
image

Preview deployed: https://fankserver.com/calendar

### CODE REVIEW

**Comment by Fank** [comment]
@pitaj could you please review the changes i made while i try to complete it?

**Comment by pitaj** [comment]
I'm really wondering why these diffs are messed up so bad. Any idea why?

**Comment by Fank** [comment]
Maybe whitespace changes i did some, try to hide whitespace changes. I think its ?w=1

File: package.json [hunk 1]
```diff
@@ -58,7 +58,10 @@
     "core-js": "^2.4.1",
     "eonasdan-bootstrap-datetimepicker": "^4.17.47",
     "fullcalendar": "^3.4.0",
+    "ical.js": "^1.2.2",
     "moment": "^2.18.1",
+    "request": "^2.81.0",
```
Comment (pitaj) [inline package.json]: Don't think we need `request` as well as `request-promise`. IIRC request promise includes request itself.

Comment (Fank) [inline package.json]: can you show me an example of it? didn't found a request example or something like that.

File: src/lib/templates.js [hunk 2]
```diff
@@ -118,19 +122,25 @@ const eventTemplate = ({ event, isEmail, uid }) => {
...
     ` : ''}
-    ${event.description.length ? `
+    ${event.description ? `
     <div class="plugin-calendar-event-description">
       <i class="fa fa-info-circle" aria-hidden="true"></i>
       <div>${event.description}</div>
     </div>
     ` : ''}
-    ${event.reminders.length ? `
+    ${!event.external && event.reminders ? `
```
Comment (Fank) [inline src/lib/templates.js]: It must be `event.reminders.length` because its an array.

File: package.json (hunk 1 above)
Comment (pitaj) [inline package.json]: I think you just don't need to have `request` as a dependency.

File: src/lib/templates.js (hunk 2 above)
Comment (pitaj) [inline src/lib/templates.js]: Is this a reminder for you to re-add `.length` for these? I'm just a little confused because empty arrays are truthy.

Comment (Fank) [inline src/lib/templates.js]: Yea :D didn't have time so i added a comment to change it later.

File: package.json (hunk 1 above)
Comment (Fank) [inline package.json]: Its a linked via request https://github.com/request/request#promises--asyncawait

Comment (pitaj) [inline package.json]: Ah actually it appears that it is a [required peer dependency](https://github.com/request/request-promise#migration-from-v2-to-v3)

File: public/admin.js [hunk 3]
```diff
@@ -1,15 +1,56 @@
 /* global $, app, define */
 /* eslint no-var: 0, prefer-arrow-callback: 0, func-names: 0 */
 
-define('admin/plugins/calendar', [], function () {
-  $('#save').on('click', function () {
+define('admin/plugins/calendar', [], function() {
```
Comment (pitaj) [inline public/admin.js]: Please don't remove these spaces. It's part of the style guide I use.

**Comment by Fank** [comment]
@pitaj I think i'm done for now, tested a lot of stuff and looks like everything works.
There are some things which should be improved for later:
- Rename "ical" and "externalevent" to something more global, to support jcal, ....
- Add styles to show calendar entries in different color foreach imported calendar.
- TBD

File: public/templates/calendar.tpl [hunk 4]
```diff
@@ -28,7 +28,7 @@ role="dialog" aria-labelledby="plugin-calendar-cal-event-display-title">
         <a type="button" href="{relative_path}/post/{eventData.pid}" itemprop="url" class="btn btn-primary gotopost" data-dismiss="modal">
           [[calendar:go_to_post]]
         </a>
-        <a type="button" href="{eventData.url}" target="_self" itemprop="url" class="btn btn-primary gotourl">
+        <a type="button" href="{eventData.url}" target="_self" itemprop="url" class="btn btn-primary gotourl" style="display: none;">
```
Comment (pitaj) [inline public/templates/calendar.tpl]: Put the buttons inside an if statement instead

Comment (Fank) [inline public/templates/calendar.tpl]: I can't because we need it in the html for later.

Comment (pitaj) [inline public/templates/calendar.tpl]: Oh, of course, sorry I thought this was on a different template. 🤦‍♂️

**Comment by robjloranger** [comment]
Any interest in getting this cleaned up? I could take a look at the conflicts and test it

**Comment by Fank** [comment]
This is mine?
Amazing ...

@robjloranger i think i will not continue on this anymore, becaue i don't even use nodebb anymore, feel free to pick it and adjust it freely.
I also added you as contributor to my repository.

**Comment by robjloranger** [comment]
> This is mine?
> Amazing ...
> 
> @robjloranger i think i will not continue on this anymore, becaue i don't even use nodebb anymore, feel free to pick it and adjust it freely.
> I also added you as contributor to my repository.

Thank you, I will start this evening.

**Comment by robjloranger** [comment]
@pitaj would you say it is worth it to try an merge 1.0.x into these changes and go from there, or better to start fresh off of main and replicate the changes within the current codebase? it's looking like a lot has changed since 2017 :grin:

**Comment by pitaj** [comment]
Probably easiest to start fresh.

**Pull request closed without merging**